File: app/src/test/java/com/example/mike/mp3player/service/MediaPlaybackServiceTest.java
package com.example.mike.mp3player.service;

import android.os.Bundle;
import android.support.v4.media.MediaBrowserCompat;
import android.support.v4.media.session.MediaSessionCompat;

import androidx.media.MediaBrowserServiceCompat;

import org.apache.commons.lang3.reflect.FieldUtils;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;

import java.util.ArrayList;
import java.util.List;


import static com.example.mike.mp3player.commons.Constants.PACKAGE_NAME;

import static com.example.mike.mp3player.commons.Constants.REQUEST_OBJECT;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

@RunWith(RobolectricTestRunner.class)
public class MediaPlaybackServiceTest {
    private static final String ACCEPTED_MEDIA_ROOT_ID = "";
    private static final String REJECTED_MEDIA_ROOT_ID = "";
    /** object to test*/
    private TestMediaPlaybackServiceInjector mediaPlaybackService;

    @Before
    public void setup() {
        this.mediaPlaybackService = Robolectric.buildService(TestMediaPlaybackServiceInjector.class).create().get();
    }

    @Test
    public void testGetRootValid() {
        MediaBrowserServiceCompat.BrowserRoot result =
                mediaPlaybackService.onGetRoot(PACKAGE_NAME, 0, null);
        assertNotNull(result);
        assertEquals(ACCEPTED_MEDIA_ROOT_ID, result.getRootId());
    }

    @Test
    public void testGetRootInvalid() {
        final String badPackageName = "bad.package.name";
        MediaBrowserServiceCompat.BrowserRoot result =
                mediaPlaybackService.onGetRoot(badPackageName, 0, null);
        assertNotNull(result);
        assertEquals(REJECTED_MEDIA_ROOT_ID, result.getRootId());
    }

    @Test
    public void testOnLoadChildrenNullExtras() {
        final String parentId = "parentId";
        MediaBrowserServiceCompat.Result<List<MediaBrowserCompat.MediaItem>> result = mock(MediaBrowserServiceCompat.Result.class);
        mediaPlaybackService.onLoadChildren(parentId, result, null);
        verify(result, times(1)).sendResult(null);
    }

    @Test
    public void testOnLoadChildrenRejectedMediaId() {
        MediaBrowserServiceCompat.Result<List<MediaBrowserCompat.MediaItem>> result = mock(MediaBrowserServiceCompat.Result.class);
        mediaPlaybackService.onLoadChildren(REJECTED_MEDIA_ROOT_ID, result);
        verify(result, times(1)).sendResult(null);
    }

    @Test
    public void testOnLoadChildrenNullLibraryRequest() {
        final String parentId = "abcd";
        MediaBrowserServiceCompat.Result<List<MediaBrowserCompat.MediaItem>> result = mock(MediaBrowserServiceCompat.Result.class);
        Bundle extras = new Bundle();
        extras.putString(REQUEST_OBJECT, null);
        mediaPlaybackService.onLoadChildren(parentId, result, extras);
        verify(result, times(1)).sendResult(null);
    }

    @Test
    public void testOnLoadChildrenValidLibraryRequest() {
        final String parentId = "abcd";
        MediaBrowserServiceCompat.Result<List<MediaBrowserCompat.MediaItem>> result = mock(MediaBrowserServiceCompat.Result.class);
        mediaPlaybackService.onLoadChildren(parentId, result);
        verify(result, times(1)).sendResult(any(ArrayList.class));
    }

    @Test
    public void testOnDestroy() throws IllegalAccessException {
        MediaSessionCompat mediaSessionSpy = spy(mediaPlaybackService.getMediaSession());
        FieldUtils.writeField(mediaPlaybackService, "mediaSession", mediaSessionSpy, true);
        mediaPlaybackService.onDestroy();
        verify(mediaSessionSpy, times(1)).release();
    }

}

File: service/src/androidTestFullDebug/java/com/github/goldy1992/mp3player/service/MediaPlaybackServiceTest.kt
package com.github.goldy1992.mp3player.service

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.os.IBinder
import android.util.Log
import androidx.media3.session.IMediaSessionService
import androidx.media3.session.MediaBrowser
import androidx.media3.session.MediaLibraryService
import androidx.media3.session.MediaSessionService
import androidx.media3.session.SessionToken
import androidx.test.core.app.ApplicationProvider
import androidx.test.rule.ServiceTestRule
import com.github.goldy1992.mp3player.service.dagger.modules.service.ContentManagerModule
import com.github.goldy1992.mp3player.service.dagger.modules.service.SearchDatabaseModule
import com.github.goldy1992.mp3player.service.library.ContentManager
import com.google.common.util.concurrent.ListenableFuture
import dagger.hilt.android.testing.HiltAndroidRule
import dagger.hilt.android.testing.HiltAndroidTest
import dagger.hilt.android.testing.UninstallModules
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.StandardTestDispatcher
import kotlinx.coroutines.test.TestCoroutineScheduler
import kotlinx.coroutines.test.TestScope
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.mockito.kotlin.mock
import java.util.concurrent.TimeoutException

@OptIn(ExperimentalCoroutinesApi::class)
@UninstallModules(
    SearchDatabaseModule::class,
    ContentManagerModule::class)
@HiltAndroidTest
class MediaPlaybackServiceTest {

    private val testScheduler = TestCoroutineScheduler()

    private val dispatcher  = StandardTestDispatcher(testScheduler)
    private val testScope = TestScope(dispatcher)
    @Rule(order = 0) @JvmField
    val rule : HiltAndroidRule = HiltAndroidRule(this)
    @get:Rule(order = 1)
    val serviceRule = ServiceTestRule()



    /** object to testFullDebug */
    lateinit var mediaPlaybackService: MediaPlaybackService

    private val rootAuthenticator: RootAuthenticator = mock<RootAuthenticator>()

    private val contentManager : ContentManager = mock<ContentManager>()

    @Before
    fun setup() {
        rule.inject()
     //   mediaPlaybackService = Robolectric.setupService(MediaPlaybackService::class.java)
      //  mediaPlaybackService.player
    }

    @Test
    @Throws(TimeoutException::class)
    fun testWithBoundService() = runTest(dispatcher) {
        val context = ApplicationProvider.getApplicationContext<Context>()
        // Create the service Intent.
        val serviceIntent = Intent(
            ApplicationProvider.getApplicationContext<Context>(),
            MediaPlaybackService::class.java
        ).apply {
            this.action = MediaLibraryService.SERVICE_INTERFACE
            // Data can be passed to the service via the Intent.
            //putExtra(SEED_KEY, 42L)
        }

        // Bind the service and grab a reference to the binder.
        //val binder: IBinder = serviceRule.bindService(serviceIntent)


        // Get the reference to the service, or you can call
        // public methods on the binder directly.

      //  val service: IMediaSessionService = (binder as IMediaSessionService)

        val sessionToken = SessionToken(context, ComponentName(context, MediaPlaybackService::class.java))
        val mblf : ListenableFuture<MediaBrowser> = MediaBrowser.Builder(context, sessionToken).buildAsync()
        mblf.await()
        Log.i("", "")
        // Verify that the service is working correctly.
     //   assertThat(service.getRandomInt(), `is`(any(Int::class.java)))
    }


//    @Test
//    fun testGetRoot() {
//        val browserRoot = MediaBrowserServiceCompat.BrowserRoot(ACCEPTED_MEDIA_ROOT_ID, null)
//        val clientPackageName = "packageName"
//        val clientUid = 45
//        val rootHints: Bundle? = null
//        whenever(rootAuthenticator.authenticate(clientPackageName, clientUid, rootHints)).thenReturn(browserRoot)
//        val result = mediaPlaybackService.onGetRoot(clientPackageName, clientUid, rootHints)
//        Assert.assertNotNull(result)
//        Assert.assertEquals(ACCEPTED_MEDIA_ROOT_ID, result!!.rootId)
//
//    }

//    @Test
//    fun testOnLoadChildrenWithRejectedRootId() {
//        whenever(rootAuthenticator.rejectRootSubscription(any())).thenReturn(true)
//        val parentId = "aUniqueId"
//        val result: Result<List<MediaItem>> = mock<Result<List<MediaItem>>>()
//        mediaPlaybackService.onLoadChildren(parentId, result)
//        Shadows.shadowOf(Looper.getMainLooper()).idle()
//        verify(result, times(1)).sendResult(null)
//    }

//    @Test
//    @ExperimentalCoroutinesApi
//    fun testOnLoadChildrenWithAcceptedMediaId() = runBlockingTest {
//        val parentId = "aUniqueId"
//        val result: Result<List<MediaItem>> = mock<Result<List<MediaItem>>>()
//        val mediaItemList: List<MediaItem> = ArrayList()
//        whenever(contentManager.getChildren(any<String>())).thenReturn(mediaItemList)
//        mediaPlaybackService.onLoadChildren(parentId, result)
//        Shadows.shadowOf(Looper.getMainLooper()).idle()
//        verify(result, times(1)).sendResult(mediaItemList)
//    }
//
//    @Test
//    fun testOnLoadChildrenRejectedMediaId() {
//        whenever(rootAuthenticator.rejectRootSubscription(any())).thenReturn(true)
//        val result: Result<List<MediaItem>> = mock<Result<List<MediaItem>>>()
//        mediaPlaybackService.onLoadChildren(REJECTED_MEDIA_ROOT_ID, result)
//        Shadows.shadowOf(Looper.getMainLooper()).idle()
//        verify(result, times(1)).sendResult(null)
//    }
//
//    @Test
//    fun testOnSearch() {
//        val result : Result<MutableList<MediaItem>> = mock<Result<MutableList<MediaItem>>>()
//        val query : String = "query"
//        val extras : Bundle? = Bundle()
//        val expectedMediaItems = mock<MutableList<MediaItem>>()
//        whenever(contentManager.search(any<String>())).thenReturn(expectedMediaItems)
//        mediaPlaybackService.onSearch(query, extras, result)
//        verify(result, times(1)).sendResult(expectedMediaItems)
//    }

    @Test
    fun testOnCreate() {
        assertTrue(true)
    }

    companion object {
        private const val ACCEPTED_MEDIA_ROOT_ID = "ACCEPTED"
        private const val REJECTED_MEDIA_ROOT_ID = "REJECTED"
    }
}